Culture and Host Names

loivsen opened the pull request:
## Description

This PR adds the functionality for Culture and Hostnames Action for a Document.
Limited mock data but the UI is there. Works on the server.

May need some final adjustments after the File and Folder System PR gets merged in.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (minor updates related to the tooling or maintenance of the repository, does not impact compiled assets)

iOvergaard reviewed (changes requested):
Functionally it works. I think we need more space all around the UI:

<img width="380" alt="image" src="https://github.com/umbraco/Umbraco.CMS.Backoffice/assets/752371/a53893cd-ed8b-478c-b99e-b08f9ba655ba">

Also a note on the "use current URL": I think it should include the protocol when inserting, as in `https://localhost:xxxx` or how do you see that?

madsrasmussen commented on src/packages/documents/documents/modals/culture-and-hostnames/culture-and-hostnames-modal.element.ts:
```diff
@@ -0,0 +1,235 @@
+		const { data } = await this.#documentRepository.getCultureAndHostnames(this.#unique);
+
+		if (!data) return;
+		this._domains = data.domains.map((domain) => ({ isoCode: domain.isoCode, domainName: domain.domainName }));
```
Please make a repository specifically to work with document culture and host names. We need to split our repositories into domains to make reusability easier and to prevent extremely large repositories. You can add this repository into the culture and hostnames folder to keep the related functionality close to each other. The name of the repository could be somehing like `UmbDocumentCultureAndHostnamesRepository`.

loivsen commented on src/packages/documents/documents/modals/culture-and-hostnames/culture-and-hostnames-modal.element.ts:
```diff
@@ -0,0 +1,235 @@
+		this.modalContext?.reject();
+	}
+	#handleSave() {
+		// TODO validation before submitting
```
Current Backoffice wont let you submit the modal if there is an error, for example duplicate domains: "localhost:5173" twice.
I guess we can just prevent the submit if the API throws an error, not sure why I didn't think of that when I made it.

loivsen commented on src/packages/documents/documents/modals/culture-and-hostnames/culture-and-hostnames-modal.element.ts:
```diff
@@ -0,0 +1,235 @@
+			<umb-body-layout headline=${this.localize.term('actions_assigndomain')}>
+				<uui-box headline=${this.localize.term('assignDomain_setLanguage')}>
+					<uui-label for="select">${this.localize.term('assignDomain_language')}</uui-label>
+					<uui-select
```
That would make sense, yes! I'll add that.

loivsen commented:
> Functionally it works. I think we need more space all around the UI:
> 
> <img alt="image" width="380" src="https://private-user-images.githubusercontent.com/752371/298947626-a53893cd-ed8b-478c-b99e-b08f9ba655ba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYwODkyMTYsIm5iZiI6MTcwNjA4ODkxNiwicGF0aCI6Ii83NTIzNzEvMjk4OTQ3NjI2LWE1Mzg5M2NkLWVkOGItNDc4Yy1iOTllLWIwOGY5YmE2NTViYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEyNFQwOTM1MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZTAzMmYyNzk2N2UxNjY0YTgyYWUyYWFkYWFlMDA2YzkwY2ExYWNmMDg3ZWM0ZGJlOWYxZDFhYzlmNWFhMTNlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.dwNsYtBI3p_zSJ5xiMJpGr96xV7P_KmmXlgzV4uoQFQ">

Do you mean more spacing around the current elements (giving it more height) or perhaps a medium modal (giving it more width)?

> Also a note on the "use current URL": I think it should include the protocol when inserting, as in `https://localhost:xxxx` or how do you see that?

I did it this way just because the current Backoffice does without protocol etc. Is there a world where the user wouldn't want https included?

iOvergaard commented:
> Do you mean more spacing around the current elements (giving it more height) or perhaps a medium modal (giving it more width)?

Both, really. Ideally, we had more width to work with and a bit more padding so it doesn't look so compressed.

> > Also a note on the "use current URL": I think it should include the protocol when inserting, as in `https://localhost:xxxx` or how do you see that?
> 
> I did it this way just because the current Backoffice does without protocol etc. Is there a world where the user wouldn't want https included?

Probably there is, yeah. If it does the same as the old backoffice, then let it be as is 👍

iOvergaard reviewed (changes requested):
Looks much better. I wonder if the modal could stay open when you close? I find that I would like to see the values I just saved stay on the page, and then manually close the modal.

Functionality wise I lose the languages if I close and re-open the modal. I can only fix it by reloading, but even then it doesn't always show the languages in the "culture" dropdown before I try to add a new domain.

Take a look here:

https://github.com/umbraco/Umbraco.CMS.Backoffice/assets/752371/de909507-c1f4-49d2-bc0f-d4dce98bf654

loivsen commented:
Alright, I renamed to cancel button to a close button, and save now dont automatically close the modal.
Also think I fixed the bug with the languages not always showing. At least its been showing consistently for me now (about 10 attempts).

iOvergaard commented on src/packages/documents/documents/entity-actions/culture-and-hostnames/modal/culture-and-hostnames-modal.element.ts:
```diff
@@ -0,0 +1,249 @@
+			</uui-button>
+			<uui-popover-container id="more-options" placement="bottom-end">
+				<umb-popover-layout>
+					<uui-button label=${this.localize.term('assignDomain_addCurrent')} @click=${() => this.#onAddDomain(true)}>
+						<umb-localize key="assignDomain_addCurrent"> Add current domain </umb-localize>
+					</uui-button>
```
I dont think you need to localize this two times over as long as you leave out any whitespace in the button:

```suggestion
					<uui-button label=${this.localize.term('assignDomain_addCurrent')} @click=${() => this.#onAddDomain(true)}></uui-button>
```

loivsen replied:
Yeah, probably added the localizations a little too fast without thinking about it.

iOvergaard commented on src/packages/documents/documents/entity-actions/culture-and-hostnames/repository/culture-and-hostnames.repository.ts:
```diff
@@ -0,0 +1,51 @@
+
+		this.#init = Promise.all([
+			this.consumeContext(UMB_NOTIFICATION_CONTEXT, (instance) => {
+				this.#notificationContext = instance as UmbNotificationContext;
```
Do you know why there is a type-cast (as UmbNotificationContext) here?

loivsen replied:
Probably either VS Code acting up with the types, or I compared it to the old way of doing it. Is not needed, so Ill just remove it again :)

iOvergaard commented on src/packages/documents/documents/entity-actions/culture-and-hostnames/repository/culture-and-hostnames.repository.ts:
```diff
@@ -0,0 +1,51 @@
+
+	#dataSource: UmbDocumentCultureAndHostnamesServerDataSource;
+
+	#notificationContext?: UmbNotificationContext;
```
Consider inferring the type directly from the token:

```suggestion
	#notificationContext?: typeof UMB_NOTIFICATION_CONTEXT.TYPE;
```

iOvergaard commented on src/packages/documents/documents/entity-actions/culture-and-hostnames/repository/culture-and-hostnames.repository.ts:
```diff
@@ -0,0 +1,51 @@
+export class UmbDocumentCultureAndHostnamesRepository extends UmbBaseController implements UmbApi {
+	#init!: Promise<unknown>;
+
+	#dataSource: UmbDocumentCultureAndHostnamesServerDataSource;
```
Consider initializing the data source here directly and remove the initializer from the constructor:

```suggestion
	#dataSource = new UmbDocumentCultureAndHostnamesServerDataSource(this);
```

iOvergaard commented on src/packages/documents/documents/entity-actions/culture-and-hostnames/repository/culture-and-hostnames.repository.ts:
Consider if it's worth the trouble and if it adds something at all awaiting the `#init` promise throughout this class. You only inject the notification context and you are invoking it with a question mark throughout the code like this: `this.#notificationContext?` anyway.

loivsen replied:
This is something we do quite a lot in other places, basically just copy pasted as a blueprint to set up a repository. I guess it isn't really needed in this specific case. Looking around it seems we mostly do this for trees and stores.

iOvergaard replied:
Yeah, but this is unlikely to ever use other contexts, so I assume it would be safe without an initializer, which just makes it slower and stores more listeners in the browser memory.

loivsen replied:
I removed it now 👍

iOvergaard replied:
You should be able to remove the `#init!: Promise<unknown>;` declaration as well :)

loivsen replied:
Missed that one 🚀

iOvergaard reviewed (changes requested):
We are getting there and it's looking great. Found a few things code quality-wise you might have a look at:

iOvergaard commented on src/packages/documents/documents/entity-actions/culture-and-hostnames/culture-and-hostnames.action.ts:
```diff
@@ -0,0 +1,29 @@
+import { UmbEntityActionBase } from '@umbraco-cms/backoffice/entity-action';
+import type { UmbControllerHostElement } from '@umbraco-cms/backoffice/controller-api';
+import { UMB_MODAL_MANAGER_CONTEXT } from '@umbraco-cms/backoffice/modal';
+import type { UmbModalManagerContext } from '@umbraco-cms/backoffice/modal';
+import { UMB_CULTURE_AND_HOSTNAMES_MODAL, type UmbDocumentDetailRepository } from '@umbraco-cms/backoffice/document';
+
+export class UmbDocumentCultureAndHostnamesEntityAction extends UmbEntityActionBase<UmbDocumentDetailRepository> {
+	#modalContext?: UmbModalManagerContext;
```
```suggestion
import { UMB_CULTURE_AND_HOSTNAMES_MODAL, type UmbDocumentDetailRepository } from '@umbraco-cms/backoffice/document';

export class UmbDocumentCultureAndHostnamesEntityAction extends UmbEntityActionBase<UmbDocumentDetailRepository> {
	#modalContext?: typeof UMB_MODAL_MANAGER_CONTEXT.TYPE;
```

iOvergaard commented on src/packages/documents/documents/entity-actions/culture-and-hostnames/culture-and-hostnames.action.ts:
```diff
@@ -0,0 +1,29 @@
+		super(host, repositoryAlias, unique, entityType);
+
+		this.consumeContext(UMB_MODAL_MANAGER_CONTEXT, (instance) => {
+			this.#modalContext = instance as UmbModalManagerContext;
```
```suggestion
			this.#modalContext = instance;
```

iOvergaard reviewed (changes requested):
Just a few more preferences:
